File: educative/ShortestCommonSupersequence.java
import java.util.*;

public class ShortestCommonSupersequence 
{
	public static void main(String[] args)
	{
		Scanner scn = new Scanner(System.in);
		String str1 = scn.nextLine();
		String str2 = scn.nextLine();
		System.out.println("Shortest common super sequence length is " + SCSS(str1,str2));
	}

	public static int SCSS(String str1, String str2)
	{
		if (str1.length()==0 && str2.length()==0) return 0;

		if (str1.length()==0)
			return str2.length();

		if (str2.length()==0)
			return str1.length();

		//store scss lengths for case when no chars incldued from str2
		int[] scssLengths = new int[str2.length()+1];
		for(int i =0; i<scssLengths.length; i++)
		{
			scssLengths[i] = i;
		}

		//loop through values for rest of cases
		for(int i =0; i<str1.length(); i++)
		{
			//store 0th value corresponding to i-1
			int temp = scssLengths[0];

			//0th value for ith char = i+1 (since all chars included)
			scssLengths[0] = i+1;

			for(int j =1; j<scssLengths.length; j++)
			{
				//if i and j same, scss length = scss length i-1,j-1 + 1
				if (str1.charAt(i) == str2.charAt(j-1))
				{
					//swap temp and current value since both depend on each other's old values
					int temp2 = temp;
					temp = scssLengths[j];
					scssLengths[j] = temp2;
					scssLengths[j]++; //1 more for this char
				}
				else
				{
					temp = scssLengths[j]; // can modify , not required for scsslenghts calc
					//ith,jth not same take min(i,j-1 and i-1,j) +1
					scssLengths[j] = Math.min(scssLengths[j-1],scssLengths[j])+1;
				}
			}
		}

		return scssLengths[scssLengths.length-1];
	}
}

File: educative/ShortestCommonSupersequence.py

import sys

def shortestCommonSupersequence(str1,str2):

    if len(str1)==0 and len(str2) ==0:
        return ""

    if len(str1) == 0:
        return str2
    if len(str2) == 0:
        return str1

    temp = None
    #shortest common ss at i and j
    #if i and j are equal, then lenght = 1+ scss at i-1,j-1
    #if i and j are unequal then length 
    #min(scss(i,j-1), scss(i-1,j)) + 1

    #init values for no characters from str1 considered
    #then scss length for any j is the number of characters (index+1) in the str2
    subs = [i for i in range(len(str2)+1)] #indices from 0 to len(str2)

    #print(subs)
    for i,_ in enumerate(str1):

        temp = subs[0]  #store the i-1, 0th value in a temporary variable
        subs[0] = i+1 #scss length equals count of characters in str1 if none from str2 i.e. index +1

        for j in range(1, len(subs)):
            
            #print("i-1,j-1th : {0} i,j-1th : {1} i-1,jth : {2}".format(temp,subs[j-1],subs[j]));

            #print("considering characters {0} and {1}".format(str1[i],str2[j-1]))
            if str1[i]==str2[j-1]:
                #scss at i,j  = scss at i-1,j-1(temp) +1
                temp,subs[j] = subs[j],temp #need to swap because temp set to subs[j] and subs[j] set to temp
                subs[j] += 1 # add 1 for this letter
            else:
                temp = subs[j] #temp not involved in setting subs[j], can safely modify

                #min of scss at i,j-1 and i-1,j + 1
                subs[j] = min(subs[j-1],subs[j])+1

            #print(subs)
        
        #print("")

    #return scss length for all chars considered
    return subs[-1]
    


if __name__=="__main__":

    str1 = input()
    str2 = input()

    print("shortest common supersequence is {0}".format(shortestCommonSupersequence(str1,str2)))
